traversal/producer/GraphProducer.java

```diff
@@ -46,91 +47,102 @@
-        if (iteratorRequested.size() < parallelisation) {
-            for (int i = iteratorRequested.size(); i < parallelisation && start.hasNext(); i++) {
-                ResourceIterator<VertexMap> iterator =
-                        new GraphIterator(graphMgr, start.next(), procedure, params, filter).distinct(produced);
-                iteratorRequested.put(iterator, 0);
-            }
+    public synchronized void produce(Producer.Queue<VertexMap> queue, int request) {
+        if (isDone.get()) return;
+        else if (!isInitialised) initialise(queue);
+        distribute(queue, request);
+    }
+
+    private synchronized void initialise(Queue<VertexMap> queue) {
+        for (int i = 0; i < parallelisation && start.hasNext(); i++) {
+            ResourceIterator<VertexMap> iter =
+                    new GraphIterator(graphMgr, start.next(), procedure, params, filter).distinct(produced);
+            runningJobs.put(iter, CompletableFuture.completedFuture(null));
         }
-        if (iteratorRequested.size() == 0) {
-            done(queue);
-        } else {
-            int splitCount = (int) Math.ceil((double) count / iteratorRequested.size());
-            for (ResourceIterator<VertexMap> iterator : iteratorRequested.keySet()) {
-                assert iteratorRequested.containsKey(iterator);
-                iteratorRequested.computeIfPresent(iterator, (k, v) -> v + splitCount);
-                if (!iteratorJobs.containsKey(iterator)) {
-                    iteratorJobs.put(iterator, runAsync(() -> produceAsync(queue, iterator), forkJoinPool()));
-                }
-            }
+        isInitialised = true;
+        if (runningJobs.isEmpty()) done(queue);
+    }
+
+    private synchronized void distribute(Queue<VertexMap> queue, int request) {
+        if (isDone.get()) return;
+        int requestSplitMax = (int) Math.ceil((double) request / runningJobs.size());
+        int requestSent = 0;
+        for (ResourceIterator<VertexMap> iterator : runningJobs.keySet()) {
+            int requestSplit = Math.min(requestSplitMax, request - requestSent);
+            runningJobs.computeIfPresent(iterator, (iter, asyncJob) -> asyncJob.thenRunAsync(
+                    () -> job(queue, iter, requestSplit), forkJoinPool()
+            ));
+            requestSent += requestSplit;
+            if (requestSent == request) break;
         }
     }
 
-    private synchronized int take(ResourceIterator<VertexMap> iterator) {
-        int count = iteratorRequested.get(iterator);
-        iteratorRequested.put(iterator, 0);
-        return count;
+    private synchronized void transition(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int unfulfilled) {
+        if (iterator.hasNext()) return;
+        if (runningJobs.containsKey(iterator) && start.hasNext()) replace(queue, iterator, unfulfilled);
+        else if (!runningJobs.isEmpty() && unfulfilled > 0) distribute(queue, unfulfilled);
```

Comment:
Would changing this logic to this make the branching a bit clearer?

```
if (!runningJobs.isEmpty()) {
  if (unfulfilled > 0) distribute(queue, unfulfilled);
} else {
  done(queue);
}
```

Reply:
I've clarified this method differently now